xapian-letor/api/featurelist_internal.cc

```diff
@@ -0,0 +1,186 @@
+/** @file featurelist_internal.cc
+ * @brief Definition of Feature::Internal class.
+ */
+/* Copyright (C) 2012 Parth Gupta
+ * Copyright (C) 2016 Ayush Tomar
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian-letor/featurelist.h"
+#include "featurelist_internal.h"
+
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include "debuglog.h"
+
+using namespace std;
+using namespace Xapian;
+
+void
+FeatureList::Internal::set_database(const Xapian::Database & db)
+{
+    LOGCALL_VOID(API, "FeatureList::set_database", db);
+    featurelist_db = db;
+}
+
+void
+FeatureList::Internal::set_query(const Xapian::Query & query)
+{
+    LOGCALL_VOID(API, "FeatureList::set_query", query);
+    if (query.empty()) {
+	throw Xapian::InvalidArgumentError("Can't initialise with an empty query string");
```

Comment:
The argument isn't a query string, it's a query object.

Is an empty query object really invalid here, or just not very useful?

We don't reject all cases of query objects which can't match anything, so I'm not sure it really helps to just reject those which simplify upon construction to give `Xapian::Query::MatchNothing`.